src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.LegacyEditor/RazorWrapperFactory.cs

```diff
@@ -0,0 +1,201 @@
+        {
+            return (TResult)wrapperObj;
+        }
+
+        var wrapper = createWrapper(inner);
+        s_objectToWrapperMap.Add(inner, wrapper);
+        return wrapper;
+    }
+
+    private static ImmutableArray<TResult> WrapAll<TInner, TResult>(IReadOnlyList<TInner> list, Func<TInner, TResult> createWrapper)
+        where TInner : class
+        where TResult : class
+    {
+        var count = list.Count;
+        if (count == 0)
+        {
+            return ImmutableArray<TResult>.Empty;
+        }
+
+        using var builder = new PooledArrayBuilder<TResult>(capacity: count);
+
+        for (var i = 0; i < count; i++)
+        {
+            builder.Add(createWrapper(list[i]));
+        }
+
+        return builder.DrainToImmutable();
+    }
+
+    private static ImmutableArray<TResult> WrapAll<TInner, TResult>(IEnumerable<TInner> items, Func<TInner, TResult> createWrapper)
+        where TInner : class
+        where TResult : class
+    {
+        using var builder = new PooledArrayBuilder<TResult>();
+
+        foreach (var item in items)
+        {
+            builder.Add(createWrapper(item));
+        }
+
+        return builder.DrainToImmutable();
+    }
+
+    private static T Unwrap<T>(object obj)
+        where T : class
+        => ((Wrapper<T>)obj).Object;
+
+    private static IRazorBoundAttributeDescriptor Wrap(BoundAttributeDescriptor obj) => WrapBoundAttributeDescriptor(obj);
+    private static IRazorBoundAttributeParameterDescriptor Wrap(BoundAttributeParameterDescriptor obj) => WrapBoundAttributeParameterDescriptor(obj);
+    private static IRazorDiagnostic Wrap(RazorDiagnostic obj) => WrapDiagnostic(obj);
+    private static IRazorDocumentTracker Wrap(VisualStudioDocumentTracker obj) => WrapDocumentTracker(obj);
+    private static IRazorElementCompletionContext Wrap(ElementCompletionContext obj) => WrapElementCompletionContext(obj);
+    private static IRazorParser Wrap(VisualStudioRazorParser obj) => WrapParser(obj);
+    private static IRazorRequiredAttributeDescriptor Wrap(RequiredAttributeDescriptor obj) => WrapRequiredAttributeDescriptor(obj);
+    private static IRazorTagHelperDescriptor Wrap(TagHelperDescriptor obj) => WrapTagHelperDescriptor(obj);
+    private static IRazorTagMatchingRuleDescriptor Wrap(TagMatchingRuleDescriptor obj) => WrapTagMatchingRuleDescriptor(obj);
+
+    private static ElementCompletionContext Unwrap(IRazorElementCompletionContext obj) => Unwrap<ElementCompletionContext>(obj);
+    private static TagHelperBinding Unwrap(IRazorTagHelperBinding obj) => Unwrap<TagHelperBinding>(obj);
+    private static TagHelperDescriptor Unwrap(IRazorTagHelperDescriptor obj) => Unwrap<TagHelperDescriptor>(obj);
+    private static TagHelperDocumentContext Unwrap(IRazorTagHelperDocumentContext obj) => Unwrap<TagHelperDocumentContext>(obj);
+
+    internal static IRazorBoundAttributeDescriptor WrapBoundAttributeDescriptor(object obj)
```

Comment:
Same `object` question here. If the above was of `TInner` could this param be of type `BoundAttributeDescriptor`? it doesn't seem like we're using these on anything other than the expected types, but maybe I'm missing something here.

Reply:
The problem is that this assembly provides IVT the legacy editor and we don't want to expose any existing compiler types. Since this method is internal and expects an existing compiler type, we need to leave the parameter type `object`.

Reply:
Ahhhh. That is the piece I was missing. Thanks :)